akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -59,38 +59,82 @@ private[akka] final case class TakeWhile[T](p: T ⇒ Boolean, decider: Supervisi
+        taking = true
+        push(out, el)
+      } else
+        pull(in)
+    })
+
+    override def onResume(t: Throwable): Unit = if (!hasBeenPulled(in)) pull(in)
+    override def onRestart(t: Throwable): Unit = onResume(t)
+
+    override def onPull(): Unit = pull(in)
+    setHandlers(in, out, this)
+  }
+  override def toString = "DropWhile"
+}
+
+/**
+  * INTERNAL API
+  */
+abstract private[stream] class SupervisionedGraphStageLogic(inheritedAttributes: Attributes, shape: Shape) extends GraphStageLogic(shape) {
+  val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+  def execute(f: () ⇒ Unit): Unit =
+    try { f() } catch {
+      case NonFatal(ex) ⇒ decider(ex) match {
+        case Supervision.Stop    ⇒ onStop(ex)
+        case Supervision.Resume  ⇒ onResume(ex)
+        case Supervision.Restart ⇒ onRestart(ex)
+      }
     }
 
-  override def decide(t: Throwable): Supervision.Directive = decider(t)
+  def onResume(t: Throwable): Unit
+  def onStop(t: Throwable): Unit = failStage(t)
+  def onRestart(t: Throwable): Unit = onResume(t)
 }
 
-private[akka] object Collect {
+private[stream] object Collect {
   // Cached function that can be used with PartialFunction.applyOrElse to ensure that A) the guard is only applied once,
   // and the caller can check the returned value with Collect.notApplied to query whether the PF was applied or not.
   // Prior art: https://github.com/scala/scala/blob/v2.11.4/src/library/scala/collection/immutable/List.scala#L458
   final val NotApplied: Any ⇒ Any = _ ⇒ Collect.NotApplied
 }
 
-private[akka] final case class Collect[In, Out](pf: PartialFunction[In, Out], decider: Supervision.Decider) extends PushStage[In, Out] {
-
-  import Collect.NotApplied
+/**
+ * INTERNAL API
+ */
+private[stream] final case class Collect[In, Out](pf: PartialFunction[In, Out]) extends GraphStage[FlowShape[In, Out]] {
+  val in = Inlet[In]("Collect.in")
+  val out = Outlet[Out]("Collect.out")
+  override val shape = FlowShape(in, out)
+  override def initialAttributes: Attributes = DefaultAttributes.collect
 
-  override def onPush(elem: In, ctx: Context[Out]): SyncDirective =
-    pf.applyOrElse(elem, NotApplied) match {
-      case NotApplied             ⇒ ctx.pull()
-      case result: Out @unchecked ⇒ ctx.push(result)
-    }
+  def createLogic(inheritedAttributes: Attributes) = new SupervisionedGraphStageLogic(inheritedAttributes, shape) with InHandler with OutHandler {
+    import Collect.NotApplied
+    override def onPush(): Unit =
+      execute(() ⇒ {
+        pf.applyOrElse(grab(in), NotApplied) match {
+          case NotApplied             ⇒ pull(in)
+          case result: Out @unchecked ⇒ push(out, result)
+        }
+      })
 
-  override def decide(t: Throwable): Supervision.Directive = decider(t)
+    override def onResume(t: Throwable): Unit = if (!hasBeenPulled(in)) pull(in)
+    override def onPull(): Unit = pull(in)
```

Comment:
I still think we could have a `SupervsedPushGraphStageLogic` and not repeat this `override def onPull(): Unit = pull(in)` again and again.